src/ffi.rs

```diff
@@ -406,10 +406,19 @@ impl MatchData {
         start: usize,
         options: u32,
     ) -> Result<bool, Error> {
+        // `subject.as_ptr()` could be invalid if it was created via a type which uses lazy allocation
+        // a static empty array is created in the case where the subject length is 0
+        let (subject_ptr, subject_len) = if subject.len() > 0 {
+            (subject.as_ptr(), subject.len())
+        } else {
+            let empty_subject: [u8; 0] = [];
+            (empty_subject.as_ptr(), empty_subject.len())
+        };
+
```

Comment:
This code seems a bit over complicated to me? How about this instead:

```rust
// When the subject is empty, we use an empty slice
// with a known valid pointer. Otherwise, slices derived
// from, e.g., an empty `Vec<u8>` may not have a valid
// pointer, since creating an empty `Vec` is guaranteed
// to not allocate.
const EMPTY: &[u8] = &[];
if subject.is_empty() {
    subject = EMPTY;
}
```

Please also make sure to follow the style of the surrounding code. In particular, code and comment lines should not exceed 79 columns.